### PR TITLE
CAN and VESC Edits

### PR DESCRIPTION
**Pull request by save-buffer**
i know this isn't just CAN stuff

### CODE REVIEW

File: IO/BeagleBone/CANBusBBB.cs [hunk 1]
```diff
@@ -1,169 +1,299 @@
...
+                byte[] Payload = new byte[Frame.DataLength];
+                unsafe
+                {
+                    for (int i = 0; i < Frame.DataLength; i++) { Payload[i] = Frame.Data[i]; }
+                }
+                return new Tuple<uint, byte[]>(Frame.CANID, Payload);
+            }
+        }
+
+        public Task<Tuple<uint, byte[]>> ReadAsync()
```
Comment (Baldstrom) [inline IO/BeagleBone/CANBusBBB.cs]: Method comment?

File: IO/BeagleBone/CANBusBBB.cs [hunk 2]
```diff
@@ -1,169 +1,299 @@
...
+                }
+            });
+        }
+
+        /// <summary> Converts a DLC into a proper multiple of 4 for CAN transmission </summary>
+        /// <returns> The multipole of 4 </returns>
+        /// <param name="CanDLC"> Can dlc. </param>
+        private byte DLCToLength(byte CanDLC)
+        {
+            byte[] Dlc2Len = { 0, 1, 2, 3, 4, 5, 6, 7, 8, 12, 16, 20, 24, 32, 48, 64 };
```
Comment (Baldstrom) [inline IO/BeagleBone/CANBusBBB.cs]: Capitalization, `DLC`

File: Components/Motors/VESC.cs [hunk 3]
```diff
@@ -9,31 +9,97 @@ namespace Scarlet.Components.Motors
...
+        public sbyte TargetRPM { get; private set; } // Target RPM of the motor
 
         /// <summary> Initializes a VESC Motor controller </summary>
         /// <param name="UARTBus"> UART output to control the motor controller </param>
         /// <param name="MaxSpeed"> Limiting factor for speed (should never exceed + or - this val) </param>
         /// <param name="CANForwardID"> CAN ID of the motor controller (-1 to disable CAN forwarding) </param>
-        /// <param name="SpeedFilter"> Filter to use with MC. Good for ramp-up protection and other applications </param>
-        public VESC(IUARTBus UARTBus, float MaxSpeed, int CANForwardId = -1, IFilter<float> SpeedFilter = null)
+        /// <param name="RPMFilter"> Filter to use with MC. Good for ramp-up protection and other applications </param>
+        public VESC(IUARTBus UARTBus, float MaxSpeed, int CANForwardID = -1, IFilter<sbyte> RPMFilter = null, int ErpmPerRpm = 518, sbyte MotorMaxRpm = 60)
```
Comment (CaiB) [inline Components/Motors/VESC.cs]: `ErpmPerRpm` and `MotorMaxRpm` probably shouldn't have default values unless these are specific to the motor controller. Same for the other ctor.

File: IO/BeagleBone/BBBPinManager.cs [hunk 4]
```diff
@@ -665,16 +668,22 @@ static List<string> GenerateDeviceTree()
                     {
                         switch (Entry.Key)
                         {
-                            case BBBPin.P9_22: case BBBPin.P9_31: // 0_A
-                            case BBBPin.P9_21: case BBBPin.P9_29: // 0_B
+                            case BBBPin.P9_22:
+                            case BBBPin.P9_31: // 0_A
+                            case BBBPin.P9_21:
+                            case BBBPin.P9_29: // 0_B
```
Comment (CaiB) [inline IO/BeagleBone/BBBPinManager.cs]: Please undo this change, these lines are formatted in this way on purpose.

File: IO/BeagleBone/BBBPinManager.cs [hunk 5]
```diff
@@ -714,24 +723,28 @@ static List<string> GenerateDeviceTree()
...
-                    foreach(KeyValuePair<BBBPin, PinAssignment> Entry in I2CMappings)
+                    foreach (KeyValuePair<BBBPin, PinAssignment> Entry in I2CMappings)
                     {
-                        switch(Entry.Key)
+                        switch (Entry.Key)
                         {
-                            case BBBPin.P9_17: case BBBPin.P9_24: // 1_SCL
-                            case BBBPin.P9_18: case BBBPin.P9_26: // 1_SDA
+                            case BBBPin.P9_17:
+                            case BBBPin.P9_24: // 1_SCL
```
Comment (CaiB) [inline IO/BeagleBone/BBBPinManager.cs]: Same here.

File: IO/BeagleBone/BBBPinManager.cs [hunk 6]
```diff
@@ -766,10 +779,12 @@ static List<string> GenerateDeviceTree()
                             case BBBPin.P9_18: // 0_D1
                             case BBBPin.P9_17: // 0_CS0
                                 SPIDev0.Add(Entry.Key, Entry.Value); continue;
-                            case BBBPin.P9_31: case BBBPin.P9_42: // 1_CLK
+                            case BBBPin.P9_31:
+                            case BBBPin.P9_42: // 1_CLK
```
Comment (CaiB) [inline IO/BeagleBone/BBBPinManager.cs]: And here

File: IO/BeagleBone/BBBPinManager.cs [hunk 7]
```diff
@@ -766,10 +779,12 @@ static List<string> GenerateDeviceTree()
...
                             case BBBPin.P9_29: // 1_D0
                             case BBBPin.P9_30: // 1_D1
-                            case BBBPin.P9_20: case BBBPin.P9_28: // 1_CS0
+                            case BBBPin.P9_20:
+                            case BBBPin.P9_28: // 1_CS0
```
Comment (CaiB) [inline IO/BeagleBone/BBBPinManager.cs]: As well as here

File: IO/BeagleBone/CANBusBBB.cs [hunk 8]
```diff
@@ -1,169 +1,301 @@
...
+                    }
+                    return new Tuple<uint, byte[]>(Frame.CANID, Payload);
+                }
+            });
+        }
+
+        /// <summary> Converts a DLC into a proper multiple of 4 for CAN transmission </summary>
+        /// <returns> The multipole of 4 </returns>
+        /// <param name="CanDLC"> Can dlc. </param>
+        private byte DLCToLength(byte CanDLC)
```
Comment (CaiB) [inline IO/BeagleBone/CANBusBBB.cs]: Please add an explanation of what DLC is, as it's not obvious to someone new to using CAN

File: IO/BeagleBone/CANBusBBB.cs [hunk 9]
```diff
@@ -1,169 +1,301 @@
...
+                    {
+                        for (int i = 0; i < Frame.DataLength; i++) { Payload[i] = Frame.Data[i]; }
+                    }
+                    return new Tuple<uint, byte[]>(Frame.CANID, Payload);
+                }
+            });
+        }
+
+        /// <summary> Converts a DLC into a proper multiple of 4 for CAN transmission </summary>
+        /// <returns> The multipole of 4 </returns>
```
Comment (CaiB) [inline IO/BeagleBone/CANBusBBB.cs]: `multipole`? Didn't realize this was an antenna :P

File: IO/BeagleBone/CANBusBBB.cs [hunk 10]
```diff
@@ -1,169 +1,301 @@
...
+        /// <param name="CanDLC"> Can dlc. </param>
```
Comment (CaiB) [inline IO/BeagleBone/CANBusBBB.cs]: Unless you are explaining something about the parameter, don't bother putting this. It's a waste of space.

File: IO/BeagleBone/CANBusBBB.cs [hunk 11]
```diff
@@ -1,169 +1,301 @@
...
+        private byte DLCToLength(byte CanDLC)
+        {
+            byte[] DLC2Len = { 0, 1, 2, 3, 4, 5, 6, 7, 8, 12, 16, 20, 24, 32, 48, 64 };
+            return DLC2Len[CanDLC & 0x0F];
+        }
+
+        /// <summary> Converts a length to a CAN data length for use with DLCToLength. </summary>
+        /// <returns> The DLC. </returns>
+        /// <param name="CanLen"> Can length. </param>
```
Comment (CaiB) [inline IO/BeagleBone/CANBusBBB.cs]: Again, completely useless doc line.

File: IO/BeagleBone/CANBusBBB.cs [hunk 12]
```diff
@@ -1,169 +1,301 @@
...
+        private byte LengthToDLC(byte CanLen)
```
Comment (CaiB) [inline IO/BeagleBone/CANBusBBB.cs]: Should probably be `CANLen`

File: IO/BeagleBone/CANBusBBB.cs [hunk 13]
```diff
@@ -1,169 +1,301 @@
...
+                    12, 12, 12, 12,             /* 21 - 24 */
+                    13, 13, 13, 13, 13, 13, 13, 13,     /* 25 - 32 */
+                    14, 14, 14, 14, 14, 14, 14, 14,     /* 33 - 40 */
+                    14, 14, 14, 14, 14, 14, 14, 14,     /* 41 - 48 */
+                    15, 15, 15, 15, 15, 15, 15, 15,     /* 49 - 56 */
+                15, 15, 15, 15, 15, 15, 15, 15 /* 57 - 64 */
+            };
+            if (CanLen > 64) { return 0xF; }
+            return Len2DLC[CanLen];
+        }
```
Comment (CaiB) [inline IO/BeagleBone/CANBusBBB.cs]: These two method names still aren't clear. Below, you do `DLCToLength(LengthToDLC())`, which looks like a pointless conversion, even though it's not. Change the names to something more clear and descriptive.

File: IO/BeagleBone/CANBusBBB.cs [hunk 14]
```diff
@@ -1,169 +1,301 @@
...
-	}
+    public static class CANBBB
+    {
+        public static CANBusBBB CANBus0 { get; private set; }
+        public static CANBusBBB CANBus1 { get; private set; }
+
+        /// <summary> Prepares the given CAN buses for use. Should only be called from BeagleBone.Initialize(). </summary>
+        static internal void Initialize(bool[] EnableBuses, bool[] ExtendedCan)
+        {
+            if (EnableBuses == null || EnableBuses.Length != 2) { throw new Exception("Invalid enable array given to CANBBB.Initialize."); }
```
Comment (CaiB) [inline IO/BeagleBone/CANBusBBB.cs]: Add a check to make sure the `ExtendedCAN` array is valid as well.

File: IO/BeagleBone/CANBusBBB.cs [hunk 15]
```diff
@@ -1,169 +1,301 @@
...
-		/// <summary> Cleans up the bus object, freeing resources. </summary> 
-		public void Dispose() => close(Socket);
-	}
+    public static class CANBBB
+    {
+        public static CANBusBBB CANBus0 { get; private set; }
+        public static CANBusBBB CANBus1 { get; private set; }
+
+        /// <summary> Prepares the given CAN buses for use. Should only be called from BeagleBone.Initialize(). </summary>
+        static internal void Initialize(bool[] EnableBuses, bool[] ExtendedCan)
```
Comment (CaiB) [inline IO/BeagleBone/CANBusBBB.cs]: `ExtendedCan` -> `ExtendedCAN`

File: Components/Motors/VESC.cs [hunk 16]
```diff
@@ -9,31 +9,97 @@ namespace Scarlet.Components.Motors
...
+            this.CANForwardID = CANForwardID;
+            this.MaxRPM = Math.Abs(MaxRPM);
+            this.ERPM_PER_RPM = Math.Abs(ErpmPerRpm);
+            this.MOTOR_MAX_RPM = Math.Abs(MotorMaxRpm);
+            this.RPMFilter = RPMFilter;
+            this.SetRPMDirectly(0);
+            SetSpeedThreadFactory().Start();
+        }
+
+        public VESC(ICANBus CANBus, float MaxSpeed, uint CANID, IFilter<sbyte> RPMFilter = null, int ErpmPerRpm = 518, sbyte MotorMaxRpm = 60)
```
Comment (CaiB) [inline Components/Motors/VESC.cs]: I really like how this is one class for both communications types, I was hoping you'd do this. Nice :)

File: Components/Motors/VESC.cs [hunk 17]
```diff
@@ -9,31 +9,97 @@ namespace Scarlet.Components.Motors
...
+            this.MaxRPM = Math.Abs(MaxRPM);
+            this.CANID = CANID;
+            this.RPMFilter = RPMFilter;
+            this.ERPM_PER_RPM = Math.Abs(ErpmPerRpm);
+            this.MOTOR_MAX_RPM = Math.Abs(MotorMaxRpm);
+            this.SetRPMDirectly(0);
+            SetSpeedThreadFactory().Start();
         }
 
         public void EventTriggered(object Sender, EventArgs Event) { }
```
Comment (CaiB) [inline Components/Motors/VESC.cs]: This should be removed, as it's not used and no longer required.

File: IO/BeagleBone/CANBusBBB.cs (hunk 8 above)
Comment (save-buffer) [reply IO/BeagleBone/CANBusBBB.cs]: A DLC is a basic part of the CAN packet structure. Anyone working on this code should be familiar with what it is.

**Comment by CaiB** [comment]
This is looking good now. You can go ahead and merge, and I'll get 0.4.0 built and released tomorrow (today).
